vm-setup/roles/v1aX_integration_test/tasks/healthcheck.yml

```diff
@@ -0,0 +1,185 @@
+# Worker CAPI healthcheck and remediation controller test
+
+- name: Deploy CAPI healthcheck for worker
+  shell: kubectl apply -f "{{ CRS_PATH }}/machine-healthcheck-worker.yaml" -n "{{ NAMESPACE }}"
+
+- name: Verify that CAPI healthcheck is created
+  kubernetes.core.k8s_info:
+    api_version: cluster.x-k8s.io/v1alpha3
+    kind: MachineHealthCheck
+    namespace: "{{ NAMESPACE }}"
+  retries: 20
+  delay: 20
+  register: capi_healthcheck
+  until: (capi_healthcheck is succeeded) and
+         (capi_healthcheck.resources | length == 1)
+
+- name: Fetch the target cluster kubeconfig
+  k8s_info:
+    kind: secrets
+    name: "{{ CLUSTER_NAME }}-kubeconfig"
+    namespace: "{{ NAMESPACE }}"
+  register: metal3_kubeconfig
+
+- name: Decode and save cluster kubeconfig
+  copy:
+    content: "{{ metal3_kubeconfig.resources[0].data.value | b64decode }}"
+    dest: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+
+- name: Install Calico to workflow cluster
+  shell: kubectl --kubeconfig /tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml apply -f https://docs.projectcalico.org/v3.20/manifests/calico.yaml
+
+- name: Pause for 2 minutes. Wait for Nodes to get ready state
+  pause:
+    minutes: 1
+
+- name: Register worker nodes
+  kubernetes.core.k8s_info:
+    api_version: v1
+    kind: Node
+    label_selectors:
+      - "!node-role.kubernetes.io/control-plane"
+    kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+  register: worker_nodes
+
+- name: Set worker IP address
+  set_fact:
+    worker_ip: "{{worker_nodes.resources[0].status.addresses[0].address}}"
+
+- name: Stop kubelet on worker node
+  ansible.builtin.shell: | 
+    ssh -o "StrictHostKeyChecking no" metal3@{{ worker_ip }} /bin/bash <<'EOT'
+    echo "These commands will be run on: $( sudo su )"
+    echo "These commands will be run on: $( systemctl stop kubelet )"
+    exit 0
+    EOT
+  ignore_errors: yes
+
+- name: Wait until CAPI machine healthcheck detects the unhealthy machine
+  kubernetes.core.k8s_info:
+    api_version: cluster.x-k8s.io/v1alpha3
+    kind: MachineHealthCheck
+    namespace: "{{ NAMESPACE }}"
+  retries: 120
+  delay: 20
+  register: capi_healthcheck
+  until: (capi_healthcheck.resources[0].status.currentHealthy | default(0) == 0)
```

Comment:
The left hand side defaults to 0, which matches the right hand side. So I suspect if there's an error or a problem with the left hand side, this task will pass regardless.

Reply:
Resoved